stages/org.osbuild.ostree.passwd

```diff
@@ -0,0 +1,121 @@
+/usr/lib/passwd, merge them and store the result into /etc/passwd in the
+buildroot. Do the same for /etc/group file.
+
+The use case for this stage is when one wants to preserve UIDs and GIDs
+which might change when the system is build from scratch. Creating these
+files before any RPMs (or other packages) are installed will prevent changes
+in UIDs and GIDs.
+"""
+
+import os
+import shutil
+import sys
+import subprocess
+
+import osbuild.api
+
+
+SCHEMA_2 = """
+"options": {
+  "additionalProperties": false
+},
+"inputs": {
+  "type": "object",
+  "additionalProperties": false,
+  "required": ["commits"],
+  "properties": {
+    "commits": {
+      "type": "object",
+      "additionalProperties": true
+    }
+  }
+}
+"""
+
+
+def passwd_lines_to_dict(lines):
+    """Take a list of passwd lines and produce a "name": "line" dictionary"""
+    res = dict()
+    for line in lines:
+        res[line.split(':')[0]] = line
+
+    return res
+
+
+def merge_passwd(primary, secondary):
+    """Merge lines from two passwd files
+
+    The lines from the "primary" file take precedence over the lines from
+    the "secondary" file meaning that if a user already exists in the first
+    one, the line will not be appended.
+
+    The result is stored in the primary file.
+    """
+    new_passwd_lines = []
+    with open(primary, "r") as p:
+        new_passwd_dict = passwd_lines_to_dict(p.readlines())
+        with open(secondary, "r") as s:
+            secondary_passwd_dict = passwd_lines_to_dict(s.readlines())
+            for name, passwd_line in secondary_passwd_dict.items():
+                if name not in new_passwd_dict:
+                    new_passwd_dict[name] = passwd_line
+
+            new_passwd_lines = list(new_passwd_dict.values())
+
+    with open(primary, "w") as p:
+        p.writelines(new_passwd_lines)
+
+
+def ostree(*args, _input=None, **kwargs):
+    args = list(args) + [f'--{k}={v}' for k, v in kwargs.items()]
+    print("ostree " + " ".join(args), file=sys.stderr)
+    subprocess.run(["ostree"] + args,
+                   encoding="utf-8",
+                   stdout=sys.stderr,
+                   input=_input,
+                   check=True)
+
+
+def parse_input(inputs):
+    commits = inputs["commits"]
+    data = commits["data"]
+    refs = data["refs"]
+    assert refs, "Need at least one commit"
+    return commits["path"], data["refs"]
+
+
+# pylint: disable=too-many-statements
+def main(tree, inputs, _options):
+    source_root, commits = parse_input(inputs)
+
+    os.makedirs(os.path.join(tree, "etc"), exist_ok=True)
+    for commit, data in commits.items():
+        ref = data.get("path", commit).lstrip("/")
+        checkout_root = os.path.join(source_root, ref)
+
+        # Merge /usr/etc/passwd with /usr/lib/passwd from the checkout and store it in the buildroot
+        # Entries in /usr/etc/passwd have a precedence.
+        buildroot_etc_passwd = os.path.join(tree, "etc/passwd")
```

Comment:
`buildroot` might be slightly misleading, since we normally use that term for the container that we use to run the stages in. `treeroot`? or just no prefix?

Reply:
"fixed" by not using the variable at all :-)